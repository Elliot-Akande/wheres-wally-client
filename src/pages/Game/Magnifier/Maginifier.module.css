.magnifier {
  position: absolute;
  pointer-events: none;

  display: flex;
  justify-content: center;

  border-radius: 50px;
  border: 2px solid var(--accent);
  box-shadow: 0 0 3px 3px var(--neutral-900);
  background-repeat: no-repeat;
  transition: opacity 0.15s;
}
