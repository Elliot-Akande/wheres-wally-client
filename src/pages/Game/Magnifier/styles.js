import { styled } from "styled-components";

export const StyledMagnifier = styled.div.attrs(
  ({
    $magnifierHeight,
    $magnifierWidth,
    $coords,
    $imageUrl,
    $imageDimensions,
    $zoomLevel,
    $hidden,
  }) => ({
    style: {
      height: `${$magnifierHeight}px`,
      width: `${$magnifierWidth}px`,
      top: `${$coords.y - $magnifierHeight / 2}px`,
      left: `${$coords.x - $magnifierWidth / 2}px`,

      backgroundImage: `url('${$imageUrl}')`,
      backgroundSize: `${$imageDimensions.width * $zoomLevel}px ${
        $imageDimensions.height * $zoomLevel
      }px`,
      backgroundPositionX: `${-$coords.x * $zoomLevel + $magnifierWidth / 2}px`,
      backgroundPositionY: `${
        -$coords.y * $zoomLevel + $magnifierHeight / 2
      }px`,
      opacity: $hidden ? 0 : 1,
    },
  })
)`
  position: absolute;
  pointer-events: none;

  display: flex;
  justify-content: center;

  border-radius: 50px;
  border: 2px solid var(--accent);
  box-shadow: 0 0 3px 3px var(--neutral-900);
  background-repeat: no-repeat;
  transition: opacity 0.15s;
`;
