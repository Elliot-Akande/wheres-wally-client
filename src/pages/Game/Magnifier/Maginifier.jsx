import SelectionMenu from "../SelectionMenu/SelectionMenu";
import styles from "./Maginifier.module.css";

const Magnifier = ({
  clickedCoords,
  hoverCoords,
  magnifierBehaviour,
  characters,
  checkAnswer,
  imageUrl,
  imageDimensions,
}) => {
  const zoomLevel = 1.5;
  const magnifierHeight = 80;
  const magnifierWidth = 80;

  const coords = magnifierBehaviour === "clicked" ? clickedCoords : hoverCoords;
  const menuDirection = (() => {
    if (magnifierBehaviour !== "clicked") return;

    if (clickedCoords.x < imageDimensions.width / 2) {
      return "left";
    }
    return "right";
  })();

  const calculatedStyles = {
    height: `${magnifierHeight}px`,
    width: `${magnifierWidth}px`,
    top: `${coords.y - magnifierHeight / 2}px`,
    left: `${coords.x - magnifierWidth / 2}px`,

    backgroundImage: `url('${imageUrl}')`,
    backgroundSize: `${imageDimensions.width * zoomLevel}px ${
      imageDimensions.height * zoomLevel
    }px`,
    backgroundPositionX: `${-coords.x * zoomLevel + magnifierWidth / 2}px`,
    backgroundPositionY: `${-coords.y * zoomLevel + magnifierHeight / 2}px`,
    opacity: magnifierBehaviour === "hidden" ? 0 : 1,
  };

  return (
    <>
      <div className={styles.magnifier} style={calculatedStyles}>
        {magnifierBehaviour === "clicked" && (
          <SelectionMenu
            checkAnswer={checkAnswer}
            characters={characters}
            menuDirection={menuDirection}
          />
        )}
      </div>
    </>
  );
};

export default Magnifier;
