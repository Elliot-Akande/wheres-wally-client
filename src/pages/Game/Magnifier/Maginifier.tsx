import SelectionMenu from "../SelectionMenu/SelectionMenu";
import { StyledMagnifier } from "./styles";

interface Character {
  name: string;
  imageUrl: string;
}

interface MagnifierProps {
  clickedCoords: {
    x: number;
    y: number;
  };
  hoverCoords: {
    x: number;
    y: number;
  };
  magnifierBehaviour: "hidden" | "hover" | "clicked";
  characters: Character[];
  checkAnswer: (answer: string) => Promise<void>;
  imageUrl: string;
  imageDimensions: {
    width: number;
    height: number;
  };
}

const Magnifier = ({
  clickedCoords,
  hoverCoords,
  magnifierBehaviour,
  characters,
  checkAnswer,
  imageUrl,
  imageDimensions,
}: MagnifierProps) => {
  const magnifierProps = {
    $magnifierHeight: 80,
    $magnifierWidth: 80,
    $zoomLevel: 1.5,
    $coords: magnifierBehaviour === "clicked" ? clickedCoords : hoverCoords,
    $hidden: magnifierBehaviour === "hidden",
    $imageUrl: imageUrl,
    $imageDimensions: imageDimensions,
  };

  const getMenuDirection = (): "left" | "right" => {
    if (clickedCoords.x < imageDimensions.width / 2) {
      return "right";
    }
    return "left";
  };

  return (
    <>
      <StyledMagnifier {...magnifierProps}>
        {magnifierBehaviour === "clicked" && (
          <SelectionMenu
            checkAnswer={checkAnswer}
            characters={characters}
            menuDirection={getMenuDirection()}
          />
        )}
      </StyledMagnifier>
    </>
  );
};

export default Magnifier;
