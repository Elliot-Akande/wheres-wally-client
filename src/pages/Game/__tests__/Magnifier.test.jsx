import { render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import Magnifier from "../Magnifier/Maginifier";

vi.mock("../SelectionMenu/SelectionMenu.jsx", () => ({
  default: () => <div>Selection Menu</div>,
}));

const mockProps = {
  clickedCoords: { x: 50, y: 50 },
  hoverCoords: { x: 100, y: 100 },
  magnifierBehaviour: "hidden",
  characters: [{ name: "test", imageUrl: "test.jpg" }],
  checkAnswer: vi.fn(),
  imageUrl: "test.jpg",
  imageDimensions: { width: 200, height: 200 },
};

describe("Magnifier", () => {
  it("renders without crashing", () => {
    const { container } = render(<Magnifier {...mockProps} />);
  });

  it('displays the SelectionMenu when magnifierBehaviour is "clicked"', () => {
    const { container } = render(
      <Magnifier {...mockProps} magnifierBehaviour="clicked" />
    );
    expect(container).toMatchSnapshot();
  });

  it('does not display the SelectionMenu when magnifierBehaviour is not "clicked"', () => {
    const { container } = render(
      <Magnifier {...mockProps} magnifierBehaviour="hover" />
    );
    expect(container).toMatchSnapshot();
    screen.debug();
  });
});
