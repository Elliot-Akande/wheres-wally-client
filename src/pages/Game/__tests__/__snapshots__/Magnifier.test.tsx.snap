// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Magnifier > displays the SelectionMenu when magnifierBehaviour is "clicked" 1`] = `
<div>
  <div
    class="sc-beySPh eaciJa"
    style="height: 80px; width: 80px; top: 10px; left: 10px; background-image: url(test.jpg); background-size: 300px 300px; background-position-x: -35px; background-position-y: -35px; opacity: 1;"
  >
    <div>
      Selection Menu
    </div>
  </div>
</div>
`;

exports[`Magnifier > does not display the SelectionMenu when magnifierBehaviour is not "clicked" 1`] = `
<div>
  <div
    class="sc-beySPh eaciJa"
    style="height: 80px; width: 80px; top: 60px; left: 60px; background-image: url(test.jpg); background-size: 300px 300px; background-position-x: -110px; background-position-y: -110px; opacity: 1;"
  />
</div>
`;
